### PR TITLE
Docs: Running Kubernetes in Production

### PR DESCRIPTION
**Pull request by hjacobs**
Add first version of "Running Kubernetes in Production" with some of our learnings.

### CODE REVIEW

**Comment by szuecs** [comment]
@hjacobs You have 2 PRs with setting limits: cluster/manifests/default-limits/limits.yaml 

I am fine with this one

:+1:

**Comment by hjacobs** [comment]
@szuecs yeah, right, my fault --- let's merge #279 first.

**Comment by szuecs** [comment]
done @hjacobs

File: docs/admin-guide/kubernetes-in-production.rst [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+Graceful Pod Termination
+========================
+
+Kubernetes will cause service disruptions on pod terminations by default as applications and configuration need to be prepared for graceful shutdown. By default, pods receive the TERM signal and will be killed 30s later with KILL. Kubernetes expects the container to handle the TERM signal and change the readinessProbe_ to “fail”. This can be achieved by changing an in-memory value which instructs the health endpoint to return a non-200 status code. Sadly the default probe settings (period 10s, failure threshold 3) prevent an in-time change of the “ready” state. To achieve graceful pod shutdown..
+
+* ..the container needs to switch its health endpoint to non-200 status code when receiving the TERM signal
+* ..the readinessProbe settings need to be tweaked to react within the grace period (30s by default), e.g. by setting failure threshold to 1 or 2 (while keeping the default period of 10s)
+
+TODO: example and link existing blog posts by others
```
Comment (mikkeloscar) [inline docs/admin-guide/kubernetes-in-production.rst]: This part is not correct. I described the problem/tested it in another comment on our internal issue tracker:

> So I wasn't convinced that Kubernetes still sends requests after the SIGTERM. I did some tests, and I'm pretty sure that the requests you will see after the SIGTERM are caused by keep-alive connections.
> 
> If you disabled keep-alive when receiving the SIGTERM, you will only get a few requests after (until kube-proxy updates), but the readinessProbe has no effect after the SIGTERM.
Looking at the logs presented in https://blog.risingstack.com/graceful-shutdown-node-js-kubernetes/ actually also shows this, he stops getting requests before failureThreshold * periodSecond
> 
> Here is my test program and results (logs): https://github.com/mikkeloscar/kube-sigterm-test

**Comment by hjacobs** [comment]
@mikkeloscar updated.

File: docs/admin-guide/kubernetes-in-production.rst [hunk 2]
```diff
@@ -70,27 +70,37 @@ Default resource requests and limits can be configured via the LimitRange_ resou
...
 
-Kubernetes’ assumption that application handle the TERM signal and change their health endpoint is a blocker for seamless migration from our AWS/STUPS infrastructure to Kubernetes. In STUPS, single Docker containers run directly on EC2 instances. Graceful container termination is not needed as AWS automatically deregisters EC2 instances and drains connections from the ELB on instance termination. We therefore consider solving the graceful pod termination issue in Kubernetes on the infrastructure level. This would not require any application code changes by our users (application developers).
+You find a `blog post about graceful shutdown of node.js on Kubernetes`_ and a `small test app to see the pod termination behavior`_.
```
Comment (mikkeloscar) [inline docs/admin-guide/kubernetes-in-production.rst]: I would say:

> **Here** you find a blog...

Comment (hjacobs) [inline docs/admin-guide/kubernetes-in-production.rst]: Slightly changed to "For further reading, .."

**Comment by mikkeloscar** [comment]
:+1:

**Comment by hjacobs** [comment]
:+1:

**Pull request merged**